monitoring/v1/types.go

```diff
@@ -0,0 +1,337 @@
+	//
+	// The relevant field in that resource is: spec.evaluationInterval
+	//
+	// This is represented as a Prometheus duration, e.g. 1d, 1h30m, 5m, 10s.  You
+	// can find the upstream documentation here:
+	//
+	// https://prometheus.io/docs/prometheus/latest/configuration/configuration/#duration
+	//
+	// +kubebuilder:validation:Pattern:="^(([0-9]+)y)?(([0-9]+)w)?(([0-9]+)d)?(([0-9]+)h)?(([0-9]+)m)?(([0-9]+)s)?(([0-9]+)ms)?$"
+	// +optional
+	Interval string `json:"interval,omitempty"`
+
+	// rules is a list of sequentially evaluated alerting rules.  Prometheus may
+	// process rule groups in parallel, but rules within a single group are always
+	// processed sequentially, and all rules are processed.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	Rules []Rule `json:"rules"`
+}
+
+// Rule describes an alerting rule.
+// See Prometheus documentation:
+// - https://www.prometheus.io/docs/prometheus/latest/configuration/alerting_rules
+//
+// +k8s:openapi-gen=true
+type Rule struct {
+	// alert is the name of the alert. Must be a valid label value, i.e. only
+	// contain ASCII letters, numbers, and underscores.
+	//
+	// +kubebuilder:validation:Pattern:="^[a-zA-Z_][a-zA-Z0-9_]*$"
+	// +required
+	Alert string `json:"alert"`
+
+	// expr is the PromQL expression to evaluate. Every evaluation cycle this is
+	// evaluated at the current time, and all resultant time series become pending
+	// or firing alerts.  This is most often a string representing a PromQL
+	// expression, e.g.:
+	//
+	//   mapi_current_pending_csr > mapi_max_pending_csr
+	//
+	// In rare cases this could be a simple integer, e.g. a simple "1" if the
+	// intent is to create an alert that is always firing.  This is sometimes used
+	// to create an always-firing "Watchdog" alert in order to ensure the alerting
+	// pipeline is functional.
+	//
+	// +required
+	Expr intstr.IntOrString `json:"expr"`
+
+	// for is the time period after which alerts are considered firing after first
+	// returning results.  Alerts which have not yet fired for long enough are
+	// considered pending. This is represented as a Prometheus duration, for
+	// details on the format see:
+	// - https://prometheus.io/docs/prometheus/latest/configuration/configuration/#duration
+	//
+	// +kubebuilder:validation:Pattern:="^(([0-9]+)y)?(([0-9]+)w)?(([0-9]+)d)?(([0-9]+)h)?(([0-9]+)m)?(([0-9]+)s)?(([0-9]+)ms)?$"
+	// +optional
+	For string `json:"for,omitempty"`
+
+	// labels to add or overwrite for each alert.  The results of the PromQL
+	// expression for the alert will result in an existing set of labels for the
+	// alert, after evaluating the expression, for any label specified here with
+	// the same name as a label in that set, the label here wins and overwrites
+	// the previous value.  These should typically be short identifying values
+	// that may be useful to query against.  A common example is the alert
+	// severity:
+	//
+	//   labels:
+	//     severity: warning
```

Comment:
Have you tried an oc describe on this to see how it renders? I'm concerned it's not going to render well as I don't think the spacing is preserved in the openapi schema

Reply:
The spacing is not preserved indeed. The contains `A common example is the alert severity: \n labels: severity: warning`. I suppose that should render as
``` 
common example is the alert severity: 
 labels: severity: warning
```
Wdyt?

Reply:
Prose it is.